packages/core/content-manager/server/services/utils/populate.js

```diff
@@ -7,70 +7,123 @@ const { hasDraftAndPublish, isVisibleAttribute } = strapiUtils.contentTypes;
 const { isAnyToMany } = strapiUtils.relations;
 const { PUBLISHED_AT_ATTRIBUTE } = strapiUtils.contentTypes.constants;
 
+/**
+ * Populate the model for relation
+ * @param {Object} attribute - Attribute containing a relation
+ * @param {String} attribute.relation - type of relation
+ * @param model - Model of the populated entity
+ * @param attributeName
+ * @param {Object} options - Options to apply while populating
+ * @param {Boolean} options.countMany
+ * @param {Boolean} options.countOne
+ * @returns {true|{count: true}}
+ */
+function getRelationPopulate(attribute, model, attributeName, { countMany, countOne }) {
+  const isManyRelation = isAnyToMany(attribute);
+
+  // always populate createdBy, updatedBy, localizations etc.
+  if (!isVisibleAttribute(model, attributeName)) {
+    return true;
+  }
+  if ((isManyRelation && countMany) || (!isManyRelation && countOne)) {
+    return { count: true };
+  }
+
+  return true;
+}
+
+/**
+ * Populate the model for components
+ * @param {Object} attribute - Attribute containing the components
+ * @param {String[]} attribute.components - IDs of components
+ * @param {Object} options - Options to apply while populating
+ * @param {Boolean} options.countMany
+ * @param {Boolean} options.countOne
+ * @param {Number} options.maxLevel
+ * @param {Number} level
+ * @returns {{populate: Object}}
+ */
+function getPopulateForComponents(attribute, options, level) {
+  const populatedComponents = (attribute.components || []).map((componentUID) =>
+    getDeepPopulate(componentUID, options, level + 1)
+  );
+
+  return { populate: populatedComponents.reduce(merge, {}) };
```

Comment:
I think the merge function is called with unwanted params 3rd and 4th (`index` and `original array`).

Reply:
It is yes, but they are not used as merge uses only the first two parameters.

Reply:
Ah yes, the `fp` one only takes 2 parameters. Thanks 👌